### ISSUE TITLE
RightCurly: false negative in ALONE and anonymous classes

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4058]
````
$ cat TestClass.java
public class TestClass {
  void method() {
     new Object() { @Override protected void finalize() { "".toString(); }  int b = 10; };
  }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
         <module name="RightCurly">
             <property name="option" value="alone"/>
             <property name="tokens" value="METHOD_DEF, INSTANCE_INIT"/>
         </module>
    </module>
</module>

$ java -jar checkstyle-7.6-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:3:74: '}' at column 74 should be alone on a line. [RightCurly]
Audit done.
Checkstyle ends with 1 errors.
````
74 is a position of `}` for method `finalize`
Expected: 2 violations on line 3 (column 74 and 89).
89 is a position of `}` for anonymous class;

` }  int b = 10; }` this two `}` should be violated.
Note that shouldStartLine (false, true) does not affect the logic. So the problem is in shouldBeAloneOnLine.

**Comment by ps-sp** [comment]
I am on it.

**Comment by MEZk** [comment]
@ps-sp 
It is not a simple issue and requires a lot of effort.

**Comment by ps-sp** [comment]
@MEZk Would it be alright if I still try ? I want to give it a shot. Still, if you think that my participation might slow the development down or if I amn't able to do it I shall step aside !

**Comment by MEZk** [comment]
@ps-sp 
You are welcome. Just be careful and you are expected to write good UTs as RightCurly behavior is strange when it validates anonymous classes.

**Comment by ps-sp** [comment]
@MEZk I can't find any `INSTANCE_INIT` token in the tree being traversed. The anonymous class begins with an `OBJBLOCK`. Shouldn't `OBJBLOCK` be an accepted token if `RightCurlyCheck` needs to check anonymous classes ? 
So, Would adding OBJBLOCK  to the acceptable tokens be alright ? It gives violation at column 91 that way.

**Comment by ps-sp** [comment]
@MEZk @romani 

class EnumContainerAlone {
    private enum Suit { CLUBS, HEARTS, SPADES, DIAMONDS } // ok
}

why is this ok if `option = "alone"` is being enforced.
shouldn't it be okay only when `option=ALONE_OR_SINGLELINE` ?

**Comment by MEZk** [comment]
@ps-sp 
I think that it is due to the fact that ENUM_DEF is not in the array of [acceptable tokens](https://github.com/checkstyle/checkstyle/blob/d2ddf5c6fc6b816dce7566dd5beaff5c5841f679/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L148-L148).

**Comment by ps-sp** [comment]
@MEZk yes ! but it shouldn't be acceptable though right ? There should be violation for it right ?

**Comment by MEZk** [comment]
> but it shouldn't be acceptable though right 
Why?

**Comment by ps-sp** [comment]
Because `option = "ALONE"` 
>Represents the policy that the brace must be alone on the line

**Comment by ps-sp** [comment]
@MEZk Ohh sorry ! Now I see the confusion ! I didn't mean to say that ENUM_DEF shoudn't be an acceptable token. I meant to say,  not having that right curly brace alone should be not acceptable.

**Comment by MEZk** [comment]
@ps-sp 
> Shouldn't OBJBLOCK be an accepted token if RightCurlyCheck needs to check anonymous classes ?

Let's discuss this in the scope of PR. We both need to read a code a think whether it will be a good solution.

**Comment by romani** [comment]
original idea to validate  `}` for `alone` and we did not countered necessity of this and difference with `}` for classes and methods and other tokens where it is reasonable.

Main nuance is that `}` as alone is good for type or type-member blocks (classes, enums, interfaces, methods, static inits) and expression blocks (in expressions and statements).
Difference is that expression blocks do need `;`,`,`,`)` at the end, and type-and-type-member blocks do not need ";" and can not be used in statements/expression for usage of `alone` for them is resonable.

```
    Object obj = new Object() {

    };
`--EXPR -> EXPR [2:17]
    `--LITERAL_NEW -> new [2:17]
        |--IDENT -> Object [2:21]
        |--LPAREN -> ( [2:27]
        |--ELIST -> ELIST [2:28]
        |--RPAREN -> ) [2:28]
        `--OBJBLOCK -> OBJBLOCK [2:30]
            |--LCURLY -> { [2:30]
            `--RCURLY -> } [4:4]


    int[] arr = new int[] {1,2};

`--EXPR -> EXPR [5:16]
    `--LITERAL_NEW -> new [5:16]
        |--LITERAL_INT -> int [5:20]
        |--ARRAY_DECLARATOR -> [ [5:23]
        |   `--RBRACK -> ] [5:24]
        `--ARRAY_INIT -> { [5:26]
            |--EXPR -> EXPR [5:27]
            |   `--NUM_INT -> 1 [5:27]
            |--COMMA -> , [5:28]
            |--EXPR -> EXPR [5:29]
            |   `--NUM_INT -> 2 [5:29]
            `--RCURLY -> } [5:30]


    () -> {"".equals("Hello");}

--ELIST -> ELIST [8:22]
   `--LAMBDA -> -> [8:22]
       |--LPAREN -> ( [8:19]
       |--PARAMETERS -> PARAMETERS [8:20]
       |--RPAREN -> ) [8:20]
       `--SLIST -> { [8:25]
           |--EXPR -> EXPR [8:35]
           |   `--METHOD_CALL -> ( [8:35]
           |       |--DOT -> . [8:28]
           |       |   |--STRING_LITERAL -> "" [8:26]
           |       |   `--IDENT -> equals [8:29]
           |       |--ELIST -> ELIST [8:36]
           |       |   `--EXPR -> EXPR [8:36]
           |       |       `--STRING_LITERAL -> "Hello" [8:36]
           |       `--RPAREN -> ) [8:43]
           |--SEMI -> ; [8:44]
           `--RCURLY -> } [8:45]


    void foo() {
        someMethod(() -> {"".equals("Hello");});
    }

    --METHOD_DEF -> METHOD_DEF [7:4]
     .........
       `--SLIST -> { [7:15]
           |--EXPR -> EXPR [8:18]
     .........
           |--SEMI -> ; [8:47]
           `--RCURLY -> } [9:4]

```

Should we validate `}` of expression blocks?  We can. Is it reasonable ?  unlikely. So validation will be in a state "yes we can force it" but folllowing `;`, `,`,`)` will be really weird in look. So very unlikely to be used by user.
Even in issue description, @MEZk  did not provided example of code on how code should be looking.

I think all we need in this issue is to update xdoc description of `alone` and explain for what tokens it should be used. And `}` for blocks in expressions are skipped by design. `alone` should stay doing alone, without any corrections and nuances. 

If we want to control formatting of `}` for blocks in expressions/statements we need new strategy on this.

@MEZk , @ps-sp , @rnveach , please share your thoughts on this.

**Comment by ps-sp** [comment]
@romani 
>We can. Is it reasonable ? unlikely

I feel it is a nice feature to have if we can provide it for sake of completion.

>alone should stay doing alone, without any corrections and nuances.
If we want to control formatting of } for blocks in expressions/statements we need new strategy on this

I totally agree.

>folllowing ;, ,,) will be really weird in look

I feel we should stick to format which is most commonly used in standard source code.

**Comment by romani** [comment]
> I feel it is a nice feature to have if we can provide it for sake of completion

Hmmmm, we need introduce features for a reason. now we looks like making code more complex and benefit will be 0. Right curly brace for anonymous, arrays initializations, lambdas can not be covered,   by http://checkstyle.sourceforge.net/property_types.html#rcurly "same" only.
but even in "same" it is kind of  skipped(allowed, but not enforced) from validation is cases:
```
    Thread t = new Thread(new Runnable() {
                  ...
    }); // this is OK, allowed for better code readability
```

If we introduce new token, we need to define how it will work in each mode http://checkstyle.sourceforge.net/property_types.html#rcurly even usage will be weird.
So  we can extend our documentation for each mode and define behavior for `}` as past of non-inlined blocks and inlined in statement/expression blocks. May be it better to keep them separate in different Checks to keep logic less complicated and more reliable? (the more properties/modes in Check the more bugs in it)

========

from cases of our code I see  that there could be "same line as last element of body"
```
    ignoredAnnotations = Arrays.stream(new String[] {"Test", "Before", "After",
        "BeforeClass", "AfterClass", }).collect(Collectors.toSet());
```
or "next line from body"
```
             final Class<?>[] param = new Class<?>[] {String.class, String.class,
                String.class, Attributes.class,
             };
```

But extending current check with such new modes will over complicate logic and it will be hard to understand what token to use in what mode.
What if we create new Check just for anonymous/innitializer/lambda - where "}" is in-lined part of expression, not a structure.
LAMBDA support could stay as is in RightCurlCheck, but most likely user will use new Check for it.
```
   // good
   someMethod(() -> {String.CASE_INSENSITIVE_ORDER.equals("Hello world two!");}); 
   // good
   someMethod(() -> {
        String.CASE_INSENSITIVE_ORDER.equals("Hello world two!");
    }); 
   // looks strange , for arrays we/users have code styled like this
   someMethod(() -> {
        String.CASE_INSENSITIVE_ORDER.equals("Hello world two!");}); 
   // bad
   someMethod(() -> {
        String.CASE_INSENSITIVE_ORDER.equals("Hello world two!");
        String.CASE_INSENSITIVE_ORDER.equals("Hello world two!");
    }
    ); ///  alone_or single_line should demand this !!! but it is weird, could be done by request only.
```

**Comment by rnveach** [comment]
> And } for blocks in expressions are skipped by design

My feeling is we should enforce all behavior when possible, otherwise new contributors (and even existing ones) will try to put in their own designs. This will cause either conflicting designs throughout project, or rotating designs because different people worked on the same code. We already have quite a few PRs where we tell people to revert their unnecessary changes. So we shouldn't just remove tokens without a replacement.

We have an issue to split `option` into 2 properties at #4086 and alone was brought up in there too.

I believe current discussion is about what should be allowed after `}` for expressions when it is set to alone. Is `;` ,` or `)` allowed or not? So is the issue is what comes after it, not before it. If we went with #4086 , we could add different values for `alone` so user can customize it further. `COMPLETELY_ALONE` (nothing after it) `ALONE_WITH_FRIENDS`. Names are examples only. `;` would be considered separations (and friends in this case) and `.` and such with be continuations of the current statement.

> May be it better to keep them separate in different Checks to keep logic less complicated and more reliable?

I am fine with splitting this check into 2. One for expression curlies, and one for all other curlies.

**Comment by romani** [comment]
new issue is created  - #5945.

In scope of this issue we just need to update documentation.

**Comment by romani** [comment]
@ps-sp , I can do documentation update myself, but I really want this issue to be marked as fixed by you, as you invested significant amount of time in it. Let me know what you prefer.

http://checkstyle.sourceforge.net/config_blocks.html#RightCurly
> Checks the placement of right curly braces ('}') for if-else, try-catch-finally blocks, while-loops, for-loops, method definitions, class definitions, constructor definitions, instance and static initialization blocks. 

just add extra phrase `For right curly brace of expression blocks please follow #5945.`

http://checkstyle.sourceforge.net/property_types.html#rcurly
> This property represents the policy for checking the placement of a right curly brace ('}')

should be `This property represents the policy for checking the placement of a right curly brace ('}') in blocks but not blocks of expressions. For right curly brace of expression blocks please follow #5945`

**Comment by ps-sp** [comment]
@romani will do.

**Comment by romani** [comment]
update for documentation is merged.

No functional changes were applied. There will be new Check to control right curly for anonymous classes, array initializers, lambdas.

**Comment by Sour1985** [comment]
> ````
> $ cat TestClass.java
> public class TestClass {
>   void method() {
>      new Object() { @Override protected void finalize() { "".toString(); }  int b = 10; };
>   }
> }
> 
> $ cat TestConfig.xml
> <?xml version="1.0"?>
> <!DOCTYPE module PUBLIC
>           "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
>           "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
> 
> <module name="Checker">
>     <property name="charset" value="UTF-8"/>
>     <module name="TreeWalker">
>          <module name="RightCurly">
>              <property name="option" value="alone"/>
>              <property name="tokens" value="METHOD_DEF, INSTANCE_INIT"/>
>          </module>
>     </module>
> </module>
> 
> $ java -jar checkstyle-7.6-all.jar -c TestConfig.xml TestClass.java
> Starting audit...
> [ERROR] TestClass.java:3:74: '}' at column 74 should be alone on a line. [RightCurly]
> Audit done.
> Checkstyle ends with 1 errors.
> ````
> 74 is a position of `}` for method `finalize`
> Expected: 2 violations on line 3 (column 74 and 89).
> 89 is a position of `}` for anonymous class;
> 
> ` }  int b = 10; }` this two `}` should be violated.
> Note that shouldStartLine (false, true) does not affect the logic. So the problem is in shouldBeAloneOnLine.
